# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/painting/nsDisplayList.cpp
# Commit: 00af531b9d94
# Full Hash: 00af531b9d943a7a536494ad8e9256c327383aab
# Author: Matt Woodrow <mwoodrow@mozilla.com>
# Date: 2021-05-16 09:17:48
# Regressor Bug: 1540737
# File Overlap Count: 1
# Description:
#   Bug 1540737 - Don't allow constructing an AutoBuildingDisplayList with a new frame without requiring explicit rects in the new coordinate space. r=miko
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D114675
# ==============================================================================

diff -r eeba14b5d89a -r 00af531b9d94 layout/painting/nsDisplayList.cpp
--- a/layout/painting/nsDisplayList.cpp	Sat May 15 22:38:45 2021 +0000
+++ b/layout/painting/nsDisplayList.cpp	Sat May 15 22:38:45 2021 +0000
@@ -3641,7 +3641,10 @@
   if ((drawBackgroundColor && color != NS_RGBA(0, 0, 0, 0)) ||
       aBuilder->IsForEventDelivery()) {
     if (aAutoBuildingDisplayList && !*aAutoBuildingDisplayList) {
-      aAutoBuildingDisplayList->emplace(aBuilder, aFrame);
+      nsPoint offset = aBuilder->GetCurrentFrame()->GetOffsetTo(aFrame);
+      aAutoBuildingDisplayList->emplace(aBuilder, aFrame,
+                                        aBuilder->GetVisibleRect() + offset,
+                                        aBuilder->GetDirtyRect() + offset);
     }
     Maybe<DisplayListClipState::AutoSaveRestore> clipState;
     nsRect bgColorRect = bgRect;
@@ -3721,7 +3724,10 @@
     }
 
     if (aAutoBuildingDisplayList && !*aAutoBuildingDisplayList) {
-      aAutoBuildingDisplayList->emplace(aBuilder, aFrame);
+      nsPoint offset = aBuilder->GetCurrentFrame()->GetOffsetTo(aFrame);
+      aAutoBuildingDisplayList->emplace(aBuilder, aFrame,
+                                        aBuilder->GetVisibleRect() + offset,
+                                        aBuilder->GetDirtyRect() + offset);
     }
 
     if (bg->mImage.mLayers[i].mBlendMode != StyleBlend::Normal) {